### ISSUE TITLE
What to do with INTERFACEs defined in SystemContext or FormalAmpersand

### ISSUE THREAD

**Issue by hanjoosten** [AmpersandTarski/Ampersand#979]
In the current version of the SystemContext metamodel, interfaces are defined in `Interfaces.ifc` I at first assumed that it is required that those interfaces should be available to ALL prototypes that are being generated. Unfortunately this gives problems, for a user could very well define an interface with the same name. In such case, an error is thrown to that user that there are muliple interfaces with the same name. 

This is something we want to resolve in a generic way. I don't want to make an exception for SystemContext. FormalAmpersand could have interfaces defined too. 

For a short term solution, I omit to take the interfaces from the metamodel into account. We could introduce a swich to override this, if we see a need for this. 

I would like to have the opinions of @Michiel-s and @stefjoosten  about this matter.

**Comment by Michiel-s** [comment]
I agree and see why this is an issue.
There are a couple of INTERFACES (with UI) and a couple of APIs (interfaces without UI) defined in SystemContext.adl. Let me explain.

INTERFACES:
These are just there to play around and edit this part of the meta model of your application.
```
INTERFACE "Edit navigation menu" FOR Administrator
INTERFACE "Edit menu item" FOR Administrator
INTERFACE "List all interfaces" FOR Administrator
INTERFACE "Edit interface" FOR Administrator
```

APIs:
There are needed by the prototype framework to query the menubar and roles. They are prefixed with `PF_` (as in Prototype Framework) awaiting namespace support in Ampersand language #850 
No frontend/UI stuff is generated so a &-user won't notice them. The prefixing prevents unfortunate naming collision.
```
API "PF_MenuItems" FOR SYSTEM
API "PF_AllRoles" FOR SYSTEM
```

As @RieksJ also points out in https://github.com/AmpersandTarski/Ampersand/issues/957#issuecomment-491590795 the administrator interfaces shouldn't be in SystemContext given its purpose to only specify what's really necessary to have a working prototype. I agree with this point of view.
However, I'll forse some kind of Admin dashboard for prototypes that help you setup and configure and maintain your application after deployment. In the spirit of Ampersand we of course define such functionality in an Ampersand script.

I propose to do the following:
1. Keep the INTERFACES but assign them to a special role named ADMIN. And give interfaces for this role a special place in the prototype framework to prevent them from showing up in the navbar.
2. Keep the APIs in there, because they are needed!!
3. Give priority to (start with) implementing namespace support (#850) to solve this issue in a good way.

**Comment by stefjoosten** [comment]
Yep. Sounds like the way to go. There is no point in Han putting a lot of effort in stuff that may create more problems than we are solving, so to hold that for a moment is fine. The direction Michiel is pointing to, solve the namespace issue, should solve this. The challenge is to make this actionable. This is a challenge because we have no resources and way too much work. Here is my idea about a course of action:

Step 1. Make a metamodel for namespaces, consistent with FormalAmpersand, to expose the problem in full detail. This step is about taking our own medicine. If Ampersand helps us understand problems, than we should prototype our way to detailed understanding.
Step 2. Make demonstrators for namespaces, just like we make demonstrators for other "wicked problems" and flesh out the solution before moving this towards Haskell or the runtime framework. Let's create work items as we go.
Step 3. Do the work together, issue based.

If done right, this should be an overseeable amount of work, even for us. Does that sound like a way to go?

**Comment by hanjoosten** [comment]
Thanks for your feedback. Namespaces could help solve this problem, and I agree we should prioritize it. The &-approach Stef suggests, is the way to go. Until we have the namespaces, there is an other (in my eyes elegant) way we could deal with this problem:

Split the script into multiple files. The INTERFACE statements that we do not want in all cases should be separated into its own file. That file would INCLUDE the remainder of the script. The `addSemanticModel` part of the meatgrinder can now be a real union of the models. Based on the command/switches we will take one of both files as root of the metamodel. 

This approach is simple and elegant, and does not depend on the implementation of namespaces. There is no special behaviour required in the way metamodels are handled in the Haskell code. All metamodels are treated equal. Only diffenent root files are supplied based on switches.

**Comment by hanjoosten** [comment]
Work items for this issue:

- [ ] Split `formalAmpersand.adl` into `FAcore.adl` and `FAclient.ifc` , where client INCLUDES core
- [ ] Split `SystemContext.adl` into `SCcore.adl` and `SCclient.ifc`, where client INCLUDES core.
- [ ] make the behaviour of the `addSemanticModel` function in Haskell a proper union, so it also includes  Interfaces.
- [ ] get the metamodel stuff work again. This will be done as part of the work that is currently being done with the new commands / switches (See issue #991) 

Implementation of namespaces will be done as part of issue #850
